contracts/FundMe.sol

```diff
@@ -71,6 +71,8 @@ contract FundMe {
         ) {
             address funder = funders[funderIndex];
             addressToAmountFunded[funder] = 0; //3TODO: why is this necessary. How is this pattern called.
+            // 3ANS: This is Checks-Effects-Interactions pattern, it is used to avoid re-entrancy.
```

Comment:
What is re-entrancy? How does following this pattern protect from it?

Reply:
I will try to describe it in my own words. 
Re-entrancy is when a user calls the function for a second time before the function was executed.
If we change the amount value to zero after the function call, when a function is called the value will not be changed until the function will be executed. In this case, while the function execution is not over and the amount value is not changed, a user could call a function for a second time.
When using this pattern, amount value would already be zero before the execution of the function starts, in this way the function could not be called during the execution of the function.

Reply:
Created https://github.com/simkasss/hardhat-fund-me/issues/5 for a practical exploration task.